test/partial-affinity.js

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var allocCluster = require('./lib/test-cluster.js');
+var parallel = require('run-parallel');
+
+// TODO Dynamic kValue cases
+// TODO Dynamic relay pool cases
+// TODO Dynamic worker pool cases
+// Static cases:
+
+allocCluster.test('degenerate case for partial affinity', {
+    size: 1,
+    kValue: 1,
+    namedRemotes: ['alice'],
+    remoteConfig: {
+        'partialAffinity.enabled': true
+    },
+    seedConfig: {
+        'hyperbahn.partialAffinity': {
+            minPeersPerRelay: 1,
+            minPeersPerWorker: 1
+        }
+    }
+}, function t(cluster, assert) {
+    runScenario(cluster, assert, onCompletion);
+    function onCompletion(err) {
+        assert.end(err);
+    }
+});
+
+allocCluster.test('few relays, many workers', {
+    size: 1,
+    kValue: 1,
+    namedRemotes: ['alice', 'alice', 'alice', 'alice', 'alice'],
+    remoteConfig: {
+        'partialAffinity.enabled': true
+    },
+    seedConfig: {
+        'hyperbahn.partialAffinity': {
+            minPeersPerRelay: 1,
+            minPeersPerWorker: 1
+        }
+    }
+}, function t(cluster, assert) {
+    runScenario(cluster, assert, onCompletion);
+    function onCompletion(err) {
+        assert.end(err);
```

Comment:
The trouble is that we are not yet reaping unneeded peers, so we’ll expect any number of connections between partially connected and fully connected. The only case we can assert is invalid is “not connected”, for which there is ample evidence if the requests fail.